<a href="https://colab.research.google.com/github/jiw3026/IT_and_BA/blob/main/IT%EC%99%80_%EB%B9%84%EC%A6%88%EB%8B%88%EC%8A%A4%ED%98%81%EC%8B%A0_%EA%B3%BC%EC%A0%9C_2013009_%EB%B0%95%EC%A7%80%EC%9B%90.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
#파일 불러오기
from google.colab import files
myfile = files.upload()

Saving health_data.csv to health_data.csv


In [2]:
import io
import pandas as pd

In [3]:
data = pd.read_csv(io.BytesIO(myfile['health_data.csv']))
data.head()

기준년도  가입자일련번호  시도코드 성별코드  연령대코드(5세단위)  신장(5Cm단위)  체중(5Kg 단위)  허리둘레  시력(좌)  \
0  2020        1    36    M            9        165          60  72.1    1.2   
1  2020        2    27    F           13        150          65  81.0    0.8   
2  2020        3    11    F           12        155          55  70.0    0.6   
3  2020        4    31    M           13        160          70  90.8    1.0   
4  2020        5    41    F           12        155          50  75.2    1.5   

   시력(우)  ...  혈청크레아티닌  (혈청지오티)AST  (혈청지오티)ALT  감마지티피  흡연상태  음주여부  구강검진수검여부  \
0    1.5  ...      1.1        21.0        27.0   21.0    No    No         0   
1    0.8  ...      0.5        18.0        15.0   15.0    No    No         0   
2    0.7  ...      0.7        27.0        25.0    7.0    No    No         0   
3    1.0  ...      1.2        65.0        97.0   72.0    No    No         1   
4    1.2  ...      0.7        18.0        17.0   14.0    No    No         0   

   치아우식증유무   치석     데이터공개일자  
0      NaN  NaN  12/29/2021  
1      NaN  NaN  12/29/2021  
2      NaN  NaN  12/29/2021  
3      0.0  0.0  12/29/2021  
4      NaN  NaN  12/29/2021  

[5 rows x 31 columns]

In [4]:
#시력(독립) - 흡연(종속)
#위 주제에 관해, 세 개의 분류모델을 구성할 예정


In [4]:
#흡연상태 - 시력 간 관계 파악을 위한 분석 진행
#(1) 이를 위해 좌우 시력의 평균을 계산한 열 추가
data['total'] = data['시력(좌)'] + data['시력(우)']
data['시력평균'] = data['total'] / 2 
data.head()

기준년도  가입자일련번호  시도코드 성별코드  연령대코드(5세단위)  신장(5Cm단위)  체중(5Kg 단위)  허리둘레  시력(좌)  \
0  2020        1    36    M            9        165          60  72.1    1.2   
1  2020        2    27    F           13        150          65  81.0    0.8   
2  2020        3    11    F           12        155          55  70.0    0.6   
3  2020        4    31    M           13        160          70  90.8    1.0   
4  2020        5    41    F           12        155          50  75.2    1.5   

   시력(우)  ...  (혈청지오티)ALT  감마지티피  흡연상태  음주여부  구강검진수검여부  치아우식증유무   치석  \
0    1.5  ...        27.0   21.0    No    No         0      NaN  NaN   
1    0.8  ...        15.0   15.0    No    No         0      NaN  NaN   
2    0.7  ...        25.0    7.0    No    No         0      NaN  NaN   
3    1.0  ...        97.0   72.0    No    No         1      0.0  0.0   
4    1.2  ...        17.0   14.0    No    No         0      NaN  NaN   

      데이터공개일자  total  시력평균  
0  12/29/2021    2.7  1.35  
1  12/29/2021    1.6  0.80  
2  12/29/2021    1.3  0.65  
3  12/29/2021    2.0  1.00  
4  12/29/2021    2.7  1.35  

[5 rows x 33 columns]

In [5]:
data.head()   #기준년도, 가입자일련번호, 이름 바꾼 average 등 3개 열 삭제

기준년도  가입자일련번호  시도코드 성별코드  연령대코드(5세단위)  신장(5Cm단위)  체중(5Kg 단위)  허리둘레  시력(좌)  \
0  2020        1    36    M            9        165          60  72.1    1.2   
1  2020        2    27    F           13        150          65  81.0    0.8   
2  2020        3    11    F           12        155          55  70.0    0.6   
3  2020        4    31    M           13        160          70  90.8    1.0   
4  2020        5    41    F           12        155          50  75.2    1.5   

   시력(우)  ...  (혈청지오티)ALT  감마지티피  흡연상태  음주여부  구강검진수검여부  치아우식증유무   치석  \
0    1.5  ...        27.0   21.0    No    No         0      NaN  NaN   
1    0.8  ...        15.0   15.0    No    No         0      NaN  NaN   
2    0.7  ...        25.0    7.0    No    No         0      NaN  NaN   
3    1.0  ...        97.0   72.0    No    No         1      0.0  0.0   
4    1.2  ...        17.0   14.0    No    No         0      NaN  NaN   

      데이터공개일자  total  시력평균  
0  12/29/2021    2.7  1.35  
1  12/29/2021    1.6  0.80  
2  12/29/2021    1.3  0.65  
3  12/29/2021    2.0  1.00  
4  12/29/2021    2.7  1.35  

[5 rows x 33 columns]

In [7]:
#시력 칼럼은 1미만, 1 이상으로 변경
data['eye'] = 0


In [6]:
data.loc[data['시력평균']<1,'eye']=0
data.loc[data['시력평균']>=1,'eye']=1
data.head()  #eye에 해당 결과가 적용되었음을 확인할 수 있다. 

기준년도  가입자일련번호  시도코드 성별코드  연령대코드(5세단위)  신장(5Cm단위)  체중(5Kg 단위)  허리둘레  시력(좌)  \
0  2020        1    36    M            9        165          60  72.1    1.2   
1  2020        2    27    F           13        150          65  81.0    0.8   
2  2020        3    11    F           12        155          55  70.0    0.6   
3  2020        4    31    M           13        160          70  90.8    1.0   
4  2020        5    41    F           12        155          50  75.2    1.5   

   시력(우)  ...  감마지티피  흡연상태  음주여부  구강검진수검여부  치아우식증유무   치석     데이터공개일자  total  \
0    1.5  ...   21.0    No    No         0      NaN  NaN  12/29/2021    2.7   
1    0.8  ...   15.0    No    No         0      NaN  NaN  12/29/2021    1.6   
2    0.7  ...    7.0    No    No         0      NaN  NaN  12/29/2021    1.3   
3    1.0  ...   72.0    No    No         1      0.0  0.0  12/29/2021    2.0   
4    1.2  ...   14.0    No    No         0      NaN  NaN  12/29/2021    2.7   

   시력평균  eye  
0  1.35  1.0  
1  0.80  0.0  
2  0.65  0.0  
3  1.00  1.0  
4  1.35  1.0  

[5 rows x 34 columns]

In [7]:
#기존 시력(좌),(우) 셀은 삭제
data.head()

기준년도  가입자일련번호  시도코드 성별코드  연령대코드(5세단위)  신장(5Cm단위)  체중(5Kg 단위)  허리둘레  시력(좌)  \
0  2020        1    36    M            9        165          60  72.1    1.2   
1  2020        2    27    F           13        150          65  81.0    0.8   
2  2020        3    11    F           12        155          55  70.0    0.6   
3  2020        4    31    M           13        160          70  90.8    1.0   
4  2020        5    41    F           12        155          50  75.2    1.5   

   시력(우)  ...  감마지티피  흡연상태  음주여부  구강검진수검여부  치아우식증유무   치석     데이터공개일자  total  \
0    1.5  ...   21.0    No    No         0      NaN  NaN  12/29/2021    2.7   
1    0.8  ...   15.0    No    No         0      NaN  NaN  12/29/2021    1.6   
2    0.7  ...    7.0    No    No         0      NaN  NaN  12/29/2021    1.3   
3    1.0  ...   72.0    No    No         1      0.0  0.0  12/29/2021    2.0   
4    1.2  ...   14.0    No    No         0      NaN  NaN  12/29/2021    2.7   

   시력평균  eye  
0  1.35  1.0  
1  0.80  0.0  
2  0.65  0.0  
3  1.00  1.0  
4  1.35  1.0  

[5 rows x 34 columns]

In [8]:
data2 = pd.get_dummies(data)
data2.head()

기준년도  가입자일련번호  시도코드  연령대코드(5세단위)  신장(5Cm단위)  체중(5Kg 단위)  허리둘레  시력(좌)  \
0  2020        1    36            9        165          60  72.1    1.2   
1  2020        2    27           13        150          65  81.0    0.8   
2  2020        3    11           12        155          55  70.0    0.6   
3  2020        4    31           13        160          70  90.8    1.0   
4  2020        5    41           12        155          50  75.2    1.5   

   시력(우)  청력(좌)  ...  시력평균  eye  성별코드_F  성별코드_M  흡연상태_No  흡연상태_NotNow  \
0    1.5      1  ...  1.35  1.0       0       1        1            0   
1    0.8      1  ...  0.80  0.0       1       0        1            0   
2    0.7      1  ...  0.65  0.0       1       0        1            0   
3    1.0      1  ...  1.00  1.0       0       1        1            0   
4    1.2      1  ...  1.35  1.0       1       0        1            0   

   흡연상태_Yes  음주여부_No  음주여부_Yes  데이터공개일자_12/29/2021  
0         0        1         0                   1  
1         0        1         0                   1  
2         0        1         0                   1  
3         0        1         0                   1  
4         0        1         0                   1  

[5 rows x 38 columns]

In [9]:
data2.head()

기준년도  가입자일련번호  시도코드  연령대코드(5세단위)  신장(5Cm단위)  체중(5Kg 단위)  허리둘레  시력(좌)  \
0  2020        1    36            9        165          60  72.1    1.2   
1  2020        2    27           13        150          65  81.0    0.8   
2  2020        3    11           12        155          55  70.0    0.6   
3  2020        4    31           13        160          70  90.8    1.0   
4  2020        5    41           12        155          50  75.2    1.5   

   시력(우)  청력(좌)  ...  시력평균  eye  성별코드_F  성별코드_M  흡연상태_No  흡연상태_NotNow  \
0    1.5      1  ...  1.35  1.0       0       1        1            0   
1    0.8      1  ...  0.80  0.0       1       0        1            0   
2    0.7      1  ...  0.65  0.0       1       0        1            0   
3    1.0      1  ...  1.00  1.0       0       1        1            0   
4    1.2      1  ...  1.35  1.0       1       0        1            0   

   흡연상태_Yes  음주여부_No  음주여부_Yes  데이터공개일자_12/29/2021  
0         0        1         0                   1  
1         0        1         0                   1  
2         0        1         0                   1  
3         0        1         0                   1  
4         0        1         0                   1  

[5 rows x 38 columns]

In [18]:
data2.loc[data2['흡연상태_No']==1,'smoke']=0
data2.loc[data2['흡연상태_NotNow']==1,'smoke']=0
data2.loc[data2['흡연상태_Yes']==1,'smoke']=1
data2.tail() #smoke에 해당 결과가 적용되었음을 확인할 수 있다. 

기준년도  가입자일련번호  시도코드  연령대코드(5세단위)  신장(5Cm단위)  체중(5Kg 단위)   허리둘레  시력(좌)  \
9995  2020     9996    44           13        155          55   83.0    1.2   
9996  2020     9997    43           11        160          55   77.5    1.0   
9997  2020     9998    45           12        160          80   93.0    0.6   
9998  2020     9999    27           15        155          55   78.0    0.6   
9999  2020    10000    45           13        160          75  100.0    0.5   

      시력(우)  청력(좌)  ...  eye  성별코드_F  성별코드_M  흡연상태_No  흡연상태_NotNow  흡연상태_Yes  \
9995    1.2      1  ...  1.0       1       0        1            0         0   
9996    1.2      1  ...  1.0       1       0        1            0         0   
9997    0.6      1  ...  0.0       0       1        1            0         0   
9998    1.0      1  ...  0.0       1       0        1            0         0   
9999    0.5      1  ...  0.0       1       0        0            1         0   

      음주여부_No  음주여부_Yes  데이터공개일자_12/29/2021  smoke  
9995        0         1                   1    0.0  
9996        0         1                   1    0.0  
9997        1         0                   1    0.0  
9998        1         0                   1    0.0  
9999        0         1                   1    0.0  

[5 rows x 39 columns]

In [55]:
data2 = data2.drop(['기준년도', '가입자일련번호','데이터공개일자_12/29/2021'],axis=1)

In [56]:
data2 = data2.dropna() #결측치가 있는 경우 분류모델 형성이 안 되나, 하나하나 따져가며 제거하기엔 데이터셋이 너무 크기에 결측치 있는 행 모두 제거. 

In [57]:
from sklearn.model_selection import train_test_split

In [58]:
x_train, x_valid, y_train, y_valid = train_test_split(data2.drop('smoke', 1), data2['smoke']) #test, train 세트로 분류

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only
  """Entry point for launching an IPython kernel.


In [59]:
x_train.shape

(936, 37)

In [60]:
y_train.shape

(936,)

In [61]:
x_train.to_numpy() #dtype 오류 발생으로 인해 데이터 타입을 변경

array([[2.760e+03, 2.760e+03, 2.600e+01, ..., 0.000e+00, 1.000e+00,
        0.000e+00],
       [3.706e+03, 3.706e+03, 1.100e+01, ..., 0.000e+00, 0.000e+00,
        1.000e+00],
       [2.038e+03, 2.038e+03, 1.100e+01, ..., 0.000e+00, 1.000e+00,
        0.000e+00],
       ...,
       [2.050e+02, 2.050e+02, 2.800e+01, ..., 0.000e+00, 1.000e+00,
        0.000e+00],
       [6.166e+03, 6.166e+03, 1.100e+01, ..., 0.000e+00, 1.000e+00,
        0.000e+00],
       [5.193e+03, 5.193e+03, 4.100e+01, ..., 0.000e+00, 1.000e+00,
        0.000e+00]])

In [29]:
y_train.to_numpy()

array([0., 0., 0., ..., 0., 0., 0.])

In [47]:
x_valid.to_numpy()

array([[2.514e+03, 2.514e+03, 2.020e+03, ..., 1.000e+00, 0.000e+00,
        1.000e+00],
       [8.850e+02, 8.850e+02, 2.020e+03, ..., 1.000e+00, 0.000e+00,
        1.000e+00],
       [5.507e+03, 5.507e+03, 2.020e+03, ..., 1.000e+00, 0.000e+00,
        1.000e+00],
       ...,
       [4.954e+03, 4.954e+03, 2.020e+03, ..., 0.000e+00, 1.000e+00,
        1.000e+00],
       [3.558e+03, 3.558e+03, 2.020e+03, ..., 0.000e+00, 1.000e+00,
        1.000e+00],
       [9.346e+03, 9.346e+03, 2.020e+03, ..., 0.000e+00, 1.000e+00,
        1.000e+00]])

In [65]:
from sklearn.linear_model import LogisticRegression
lr = LogisticRegression(max_iter = 5000)
lr.fit(x_train, y_train)

LogisticRegression(max_iter=5000)

In [ ]:
pre

In [69]:
prediction  = lr.predict(x_valid)

In [67]:
prediction[:5]

array([0., 0., 1., 0., 1.])

In [68]:
(prediction ==y_valid).mean()

1.0

### 위 Logistic Regression 분류결과 1.0이 나왔다. 과적합이 의심된다. 

In [97]:
#SGD Classifier(확률적 경사하강법) 
from sklearn.linear_model import SGDClassifier

In [98]:
sgd = SGDClassifier(random_state = 5000)

In [99]:
sgd.fit(x_train, y_train)

SGDClassifier(random_state=5000)

In [100]:
prediction = sgd.predict(x_valid)

In [101]:
(prediction == y_valid).mean()

0.8210862619808307

### 위 SGD 분류 결과 약 0.821이 나왔다. 

In [102]:
#k-nn 알고리즘
from sklearn.neighbors import KNeighborsClassifier

In [103]:
knn = KNeighborsClassifier()

In [104]:
knn.fit(x_train,y_train)

KNeighborsClassifier()

In [105]:
prediction = knn.predict(x_valid)

In [106]:
(prediction==y_valid).mean()

0.7827476038338658

In [107]:
knn2 = KNeighborsClassifier(n_neighbors=9)
knn2.fit(x_train, y_train)
knn2_pred = knn2.predict(x_valid)

In [108]:
(knn2_pred == y_valid).mean() #이웃 개수를 9개로 변경하니 성능이 높아졌다.

0.8178913738019169

### k-nn에서 k를 5(default)에서 9로 높이자, 0.7827에서 0.8178로 향상된 것을 확인할 수 있었다. 다만, 위 SGD 분류기가 9-NN보다 성능이 살짝 좋다. 

In [109]:
from sklearn.svm import SVC

In [110]:
svc = SVC(random_state = 123)
svc.fit(x_train, y_train)
svc_pred = svc.predict(x_valid)

In [111]:
svc

SVC(random_state=123)

In [112]:
(svc_pred == y_valid).mean()

0.8210862619808307

### SVC(Support Vector Machine) 분류 결과 0.821로, SGD와 같은 값이 나왔다. 데이터 적합 과정 상의 문제로 유추된다. 성능 자체는 나쁘지 않다. 

In [113]:
from sklearn.tree import DecisionTreeClassifier #의사결정트리

In [114]:
dt = DecisionTreeClassifier(random_state=0)

In [115]:
dt.fit(x_train, y_train)

DecisionTreeClassifier(random_state=0)

In [116]:
dt_pred = dt.predict(x_valid)

In [117]:
(dt_pred == y_valid).mean()

1.0

### 의사결정트리 결과 1.0이 나왔다. 마찬가지로 과적합된 것으로 보인다. 

#### 담배를 피우지 말라고 하는 이유는 굉장히 많다. 살이 찌고, 폐가 나빠진다 등. 하지만 시력에 관한 정보는 많이 접하지 못해 개인적으로 궁금한 사항이었다. 이에, 실제로 관련이 있는지 확인해 보고자 흡연(경험이 있으나 끊은 경우는 No로 처리) 더미변수로 변환한 후, 분류를 진행하였다.  

#### 분류 결과, 과적합 문제가 있어 보이는 모델 2개를 제외하고는 모두 0.7 이상의 성능을 보였다. 함부로 인과관계를 도출할 수는 없겠지만 흡연이 시력에 영향을 주지 않는다고는 할 수 없는 결과라 생각한다. 